### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by danielballan** [NSLS-II/databroker-browser#15]
# Roadmap

## Status (as of start of 2019)

This project is still "alpha" code, a very experimental prototype that taught us what we needed to add to the infrastructure of databroker and bluesky to make customizable data-browsing GUIs easier to write.

The last new feature was added in May 2017, and although it was _installed_ on all beamlines at NSLS-II in 2017, it was never widely advertised or configured for use.

## Scope

To start, the GUI will have these components:

1. Search
    - Search bar with beginner-friendly filters like a calendar date-range selection tool. It should also accept arbitrarily-complex Mongo queries.
    - List of search results. There won't be room to display all the metadata, but a commonly-useful subset will be shown by default, and it will be fully customizable by beamline staff. (Provide a function takes in a Header or RunCatalog and returns a QtWidget for a given search result.)
2. Single Run Browser
    - Similar to output of BestEffortCallback
    - Uses full Qt, not just MPL as first draft of databroker-browser did. Also, allows user to change what is plotted and type in mathematical expressions (valid Python, eval-ed in a namespace with data_keys + numpy functions + whatever user adds)
    - Only shows one run (i.e. one Header) at a time
3. Export to files
     - A GUI to planned suitcase-* repos that export to a growing set of formats, starting with CSV, TIFF, Nexus, ....
     - For now, this is how users will do more sophisticated transformations/analysis or plot multiple runs together -- but moving to another piece of software. Then, the next goal is to add "Import from databroker" to popular software like Larch, pymca, etc.

## Roadmap

More infrastructural work is required before we can proceed with code in databroker-browser itself, but once that work is done, the essential features will be easy to write.

1. Do some low-level work to enable data to be represented more efficiently in "pages" (chunks along the time axis) rather than individually-boxed data-points.This is motivated by moving from system solely optimized for visualizing "live" streaming data to a system that is also optimized for quickly pulling up saved data in bulk. https://github.com/NSLS-II/event-model/pull/29
2. Finish https://github.com/NSLS-II/suitcase-csv prototype, and then working on multiple other formats like TIFF and Nexus in parallel.
3. Rework bluesky's visualization tools, which are currently optimized for live data, to also be efficient for viewing saved data. In the process, make it easier to view multiple runs at once (not in scope for the _initial_ wide release of databroker-browser, but close on the horizon). That work is in progress here: https://github.com/NSLS-II/bluesky/pull/1076
4. Rework databroker to use intake internally, which will enable databroker-browser to work on local and remote data seamlessly and to utilize dask to load only the data it actually needs. That work is in progress here: https://github.com/danielballan/intake-bluesky
5. Finally, put all the pieces together in databroker-browser. The general layout is already in place -- the search bar, the search results, the metadata viewer, some plot windows, some export buttons. But the totally-revised internals will enable a more powerful out-of-the-box user experience (no configuration required) and at the same time more flexible possibility for beamlines that want customization.

**Comment by danielballan** [comment]
## Rough Timeline

* Jan 14: Declare whatever export capability we have "good enough". Work on additional data formats in parallel, but don't hold up the main effort over it.
* Jan 21: New visualization infrastructure should have feature parity with current capabilities of BestEffortCallback, now more modular and with pagination support
* Jan 28: Reasonable draft of databroker-on-intake should be good enough for a beta GUI.
* Feb 4: Make very very experimental draft of new databroker-browser available to select testers.
